### PR TITLE
[Lang] add Matrix.dim()/shape() support

### LINKED ISSUE #1069: Shape of Tensor of Vectors/Matrices
As of now, the .shape() function works only for tensors of scalars. But it can be useful (e.g. for debugging purposes) to know the shape of the tensors of vectors/matrices.

```
# Tensor of scalars
x = ti.var(shape=(3, 2), dt=ti.f32)
# Tensor of vectors
y = ti.Vector(2, shape=(3), dt=ti.f32)

@ti.kernel
def test_tensor_of_scalars():
    # Works fine
    print(x.shape())

@ti.kernel
def test_tensor_of_vectors():
    # AttributeError: 'Matrix' object has no attribute 'shape'
    print(y.shape())
```

The same logic could be applied to ```.dim()``` as well. One potential drawback I can see from the user side is the possibility to misinterpret the indexing (```x[2][1]``` vs ```x[2, 1]```) as a consequence.

### LINKED ISSUE #1083: Different behavior of for loops in python and taichi scopes
Iterating over the tensor's values in python scope results in an infinite loop. Is it expected? Could not find the reasoning behind this in the documentation.
 
```
import taichi as ti

ti.init()

x = ti.var(shape=(3), dt=ti.f32)
x[0] = 0
x[1] = 1
x[2] = 2

@ti.kernel
def finite_loop():
    for val in x:
        print(val)

def infinite_loop():
    for val in x:
        print(val)

finite_loop()
infinite_loop()
```

Access to tensor's values by indexing also can be indefinite: ```print(x[10000000])``` works fine. As I understand, the indexing starts cycling, from the user point of view it's kinda strange, I rather expect the IndexOutOfRange exception as to be aware if there is a bug in the code.

### PR DESCRIPTION
**Pull request by archibate**
<del>... and raise an error when struct-for is used in Python-scope.</del> (non-trivial, in another PR)

Related issue = close #1069

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: python/taichi/lang/matrix.py [hunk 1]
```diff
@@ -411,6 +411,21 @@ def diag(dim, val):
...
+
+    def dim(self):
+        return self.loop_range().dim()
+
+    def __iter__(self):
+        return iter(self.loop_range())
+
+    #def data_type(self):
+    #    # XXX(@yuanming-hu): Do we need to tell users we are a matrix of dt?
+    #    return self.loop_range().data_type()
```
Comment (archibate) [inline python/taichi/lang/matrix.py:427]: Or user can easy get confused when:
```py
@ti.kernel
def clear(x: ti.template()):
  x[None] = ti.cast(x.data_type(), 0)

scalar = ti.var(dt=ti.i32, shape=())
matrix = ti.Matrix(2, 3, dt=ti.i32, shape=())

clear(scalar)  # OK
clear(matrix)  # Error: can not cast from 0 (int) to x[None] (Matrix of int)
``
Thoughts? @vkurenkov @yuanming-hu @k-ye

File: python/taichi/lang/expr.py [hunk 2]
```diff
@@ -91,6 +91,9 @@ def __setitem__(self, key, value):
                      (taichi_lang_core.get_max_num_indices() - len(key)))
         self.setter(value, *key)
 
+    def __iter__(self):
+        raise SyntaxError("Python-scope struct-for is not supported")
```
Comment (archibate) [inline python/taichi/lang/expr.py:95]: Will we consider to support Python-scope struct-for in the future?

File: python/taichi/lang/matrix.py (hunk 1 above)
Comment (yuanming-hu) [reply python/taichi/lang/matrix.py:411]: Good point! I think we should add `data_type` to `Matrix`.

**Comment by archibate** [comment]
Hello, may I merge this trivial PR on my own now?

**Pull request merged**